src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,453 @@
+          i += 1;
+        } else {
+          return thrust::make_pair(error_compoments, error_time_zone);
+        }
+      } else if (5 == i || 6 == i) {
+        if ('.' == b && 5 == i) {
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_pair(error_compoments, error_time_zone);
+          }
+          segments[i] = current_segment_value;
+          current_segment_value = 0;
+          current_segment_digits = 0;
+          i += 1;
+        } else {
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_pair(error_compoments, error_time_zone);
+          }
+          segments[i] = current_segment_value;
+          current_segment_value = 0;
+          current_segment_digits = 0;
+          i += 1;
+          tz = cudf::string_view(bytes + j, (bytes_length - j));
+          j = bytes_length - 1;
+        }
+        if (i == 6 && '.' != b) {
+          i += 1;
+        }
+      } else {
+        if (i < segments_len && (':' == b || ' ' == b)) {
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_pair(error_compoments, error_time_zone);
+          }
+          segments[i] = current_segment_value;
+          current_segment_value = 0;
+          current_segment_digits = 0;
+          i += 1;
+        } else {
+          return thrust::make_pair(error_compoments, error_time_zone);
+        }
+      }
+    } else {
+      if (6 == i) {
+        digits_milli += 1;
+      }
+      // We will truncate the nanosecond part if there are more than 6 digits, which results
+      // in loss of precision
+      if (6 != i || current_segment_digits < 6) {
+        current_segment_value = current_segment_value * 10 + parsed_value;
+      }
+      current_segment_digits += 1;
+    }
+    j += 1;
+  }
+
+  if (!is_valid_digits(i, current_segment_digits)) {
+    return thrust::make_pair(error_compoments, error_time_zone);
+  }
+  segments[i] = current_segment_value;
+
+  while (digits_milli < 6) {
+    segments[6] *= 10;
+    digits_milli += 1;
+  }
+
+  cudf::string_view timze_zone;
+  if (tz.has_value()) {
+    timze_zone = format_zone_id(tz.value());
+  } else {
+    timze_zone = default_time_zone;
+  }
+
+  segments[0] *= year_sign.value_or(1);
+  // above is translated from Spark.
+
+  // set components
+  auto components = timestamp_components{segments[0],
+                                         static_cast<int8_t>(segments[1]),
+                                         static_cast<int8_t>(segments[2]),
+                                         static_cast<int8_t>(segments[3]),
+                                         static_cast<int8_t>(segments[4]),
+                                         static_cast<int8_t>(segments[5]),
+                                         segments[6]};
+  if (just_time) {
+    components.year = components.month = components.day = -1;
+  }
+  return thrust::make_pair(components, timze_zone);
```

Comment:
If I am not wrong, the `thrust::make_pair` will trigger the copy constructor of the components. Perhaps, we can just initialize the `components` and `time_zone` in the `parse_timestamp_string_fn::operator()` and pass into this function as references or ptrs ?

Reply:
Yes, copy components(year,month,day...) is sort of waste.
Now we invoke the `to_utc_timestamp` in `parse_string_to_timestamp_us` function directly.